### PR TITLE
fix(outdated): git specificers

### PR DESCRIPTION
**Pull request by KSXGitHub**
Fixes https://github.com/pnpm/pnpm/issues/7415

### CODE REVIEW

**Comment by KSXGitHub** [comment]
@zkochan There was few things I was unable to do:

1. I don't know how to write a test that would fail without this fix. This test passes regardless of whether I include the fix:

```typescript
test('does not error on dependencies with git specifiers', async () => {
  const lockfile: Lockfile = {
    overrides: {
      'is-negative': 'github:kevva/is-negative',
    },
    importers: {
      '.': {
        dependencies: {
          'is-negative': 'github.com/kevva/is-negative/1d7e288222b53a0cab90a331f1865220ec29560c',
        },
        specifiers: {
          'is-negative': 'github:kevva/is-negative',
        },
      },
    },
    lockfileVersion: 5,
    packages: {
      'github.com/kevva/is-negative/1d7e288222b53a0cab90a331f1865220ec29560c': {
        resolution: {
          tarball: 'https://codeload.github.com/kevva/is-negative/tar.gz/1d7e288222b53a0cab90a331f1865220ec29560c',
        },
        name: 'is-negative',
        version: '2.1.0',
        engines: { node: '>=0.10.0' },
        dev: false,
      },
    },
  }
  const outdatedPkgs = await outdated({
    currentLockfile: lockfile,
    getLatestManifest,
    lockfileDir: 'project',
    manifest: {
      dependencies: {
        'is-negative': 'github:kevva/is-negative',
      },
    },
    prefix: 'project',
    wantedLockfile: lockfile,
    registries: {
      default: 'https://registry.npmjs.org/',
    },
  })
  expect(outdatedPkgs).toStrictEqual([])
})
```

2. As written in https://github.com/pnpm/pnpm/issues/7415#issuecomment-1860831986, I suspect that the issue lies elsewhere, but I was unable to figure out why only a combination of `-r` and `-i` cause such failure.

File: .changeset/nasty-games-grab.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+"@pnpm/outdated": patch
+"pnpm": patch
+---
+
+Fix error as in `outdated -i -r` with git specifiers [#7415](https://github.com/pnpm/pnpm/issues/7415).
```
Comment (zkochan) [inline .changeset/nasty-games-grab.md]: The fix is for the update command

**Comment by zkochan** [comment]
Maybe you need to test `outdatedDepsOfProjects` because the issue says that it happens with the `-r` flag.

File: .changeset/nasty-games-grab.md (hunk 1 above)
Comment (KSXGitHub) [inline .changeset/nasty-games-grab.md]:
```suggestion
Fix error as in `update -i -r` with git specifiers [#7415](https://github.com/pnpm/pnpm/issues/7415).
```

**Comment by zkochan** [comment]
In order to cover this with a test, you can add your test to the plugin-commands-outdated package. You can see how the tests leverage fixtures:

https://github.com/pnpm/pnpm/blob/main/reviewing/plugin-commands-outdated/test/recursive.ts

You can add a new fixture with overrides and git-hosted deps.

**Comment by KSXGitHub** [comment]
@zkochan This test doesn't error when I disable the fix:

```typescript

test('pnpm recursive outdated in git specifier override in root and semver in member', async () => {
  preparePackages([
    {
      location: '.',
      package: {
        pnpm: {
          overrides: {
            'is-negative': 'github:kevva/is-negative#2.1.0',
          },
        },
      },
    },
    {
      location: './project-1',
      package: {
        dependencies: {
          'is-negative': '2.1.0',
        },
      },
    },
  ])

  const { allProjects, selectedProjectsGraph } = await readProjects(process.cwd(), [])
  await install.handler({
    ...DEFAULT_OPTS,
    allProjects,
    dir: process.cwd(),
    recursive: true,
    selectedProjectsGraph,
    workspaceDir: process.cwd(),
  })

  const { output, exitCode } = await outdated.handler({
    ...DEFAULT_OPTS,
    allProjects,
    dir: process.cwd(),
    recursive: true,
    selectedProjectsGraph,
  })

  expect(exitCode).toBe(0)
  expect(output).toBe('')
})
```

**Note:** The bug only appears when BOTH `-i` and `-r` are specified. If either one is missing, pnpm runs normally.

**Comment by zkochan** [comment]
In that case I guess you can write the test for interactive recursive update.

**Comment by KSXGitHub** [comment]
A new test named `issue-7415.ts` has been created. This test would fail should [line 60-68 of `outdated.ts`](https://github.com/pnpm/pnpm/blob/2201c2afeffc3b03d2a47690f9204ab4ea1a1384/reviewing/outdated/src/outdated.ts#L60-L68) be disabled.

**Review by zkochan — APPROVED** [review]

**Pull request merged**